### PR TITLE
Implement recursive selector

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

An initial implementation of recursive selectors that offers a way to traverse them. Not covered is
validating ExploreRecursiveEdge being under an ExploreRecursive selector

# Implementation

Basically, this works in the sense that it supports recursion, including nested recursion and multiple ExploreRecursiveEdge's per ExploreRecursive

It does include one type assertion in ExploreRecursive's Explore method that enables everything to work. I wanted to check if this feels ok, and if this approach in general feels ok.

I haven't quite worked out how to validate that ExploreRecursiveEdge only appears under ExploreRecursive during selector parsing, or that every ExploreRecursive has at least one ExploreRecursiveEdge, and I wanted to verify if those validations are important to support during parse

### CODE REVIEW

**Comment by warpfork** [comment]
> It does include one type assertion in ExploreRecursive's Explore method that enables everything to work. I wanted to check if this feels ok

:+1:   Yeah, I think a type condition is a perfectly fine way to express something like this.  `Selector` is overall basically a sum type (even if it's not explicitly closed), so that feels fine to me.

> how to validate that ExploreRecursiveEdge only appears under ExploreRecursive during selector parsing, or that every ExploreRecursive has at least one ExploreRecursiveEdge, and I wanted to verify if those validations are important to support during parse

Hrm.  Dang.  Yeah, these just both kinda fail silently if written in the most obvious way, don't they.

Maybe that's actually viable, but ISTM a decent sized footgun if we don't emit errors for these.

We could make ExploreRecursiveEdge panic if ever actually called?  ExploreRecursive just eyeballs the type and then follows a branch based on it; so if anyone does call it, we've effectively detected the unparented case.  (If... rather late.  Parse time error would still be nicer.)

Can't think of a cheaper option for making sure there's at least one edge other than just doing a BFS/DFS.

I dunno if it's important for day 1, though.  Up to you.

File: traversal/selector/exploreRecursive_test.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		Wish(t, err, ShouldEqual, nil)
+		rn = n
+		rs = s
+		rs = rs.Explore(rn, PathSegmentString{S: "Side"})
+		rn, err = rn.TraverseField("Side")
+		Wish(t, rs, ShouldEqual, ExploreRecursive{subTree, ExploreRecursive{sideSelector, sideSelector, maxDepth}, maxDepth})
+		Wish(t, err, ShouldEqual, nil)
+		rs = rs.Explore(rn, PathSegmentString{S: "real"})
+		rn, err = rn.TraverseField("real")
+		Wish(t, rs, ShouldEqual, ExploreRecursive{subTree, ExploreRecursive{sideSelector, sideSelector, maxDepth - 1}, maxDepth})
```
Comment (warpfork) [inline traversal/selector/exploreRecursive_test.go]: Got a little unclear to me what these cases are trying to focus in on.  Might make sense but my eyes just kinda crossed.

Maybe with these fixtures doing some stuff with JSON (just in heredocs?  no need to bite off a whole fixture file loading thing right way) would be even more of a payoff than the earlier instances.

**Comment by hannahhoward** [comment]
@warpfork eek... so small/large problem here...

The type cast check doesn't work if ExploreRecursiveEdge sits underneath ExploreUnion...

not quite sure how to do deal with this. may need a side channel piece of state.

**Comment by warpfork** [comment]
This is heady diff.  Some partial thoughts:

- If `SelectorContext` is going to be a thing...
  - It's only during parse so I'd be nice if the type's name expressed that.
  - Not a huge fan of the vararg/slice thing.  If it's parse state, maybe this would look nicer if we did all the `ParseFoo` functions as methods on a parse state object instead?  (I like the functional take, but not so much the varargs.)

I'm trying to wrap my head around what's going to make this look best if we ported it to use codegen'd parsers later.   Does that shed any light on anything?  I'm not sure yet, but it might be a useful heuristic.

**Comment by warpfork** [comment]
Ehhh I didn't mean to *close* this :embarassed:

I think it just happened while I was doing branch cleanup and pushed `:feat/update-to-selector-spec`, which was already merged.... this PR was targeting that branch, and maybe it should be open against master now?

File: traversal/selector/exploreRecursiveEdge.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+// ParseExploreRecursiveEdge assembles a Selector
+// from a exploreRecursiveEdge selector node
+func ParseExploreRecursiveEdge(n ipld.Node, selectorContexts ...SelectorContext) (Selector, error) {
+	if n.ReprKind() != ipld.ReprKind_Map {
+		return nil, fmt.Errorf("selector spec parse rejected: selector body must be a map")
+	}
+	s := ExploreRecursiveEdge{}
+	for _, selectorContext := range selectorContexts {
+		if selectorContext.Link(s) {
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply traversal/selector/exploreRecursiveEdge.go]: The idea of "Link" is there might be future selectors added that need some kind of connection with a parent selector.

**Pull request closed without merging**